File: php/TenPinBall.php
<?php
/**
 * Created by PhpStorm.
 * User: Archit
 * Date: 9/9/2018
 * Time: 7:42 PM

10 pin bowling

1 Game : 10 Frames
1 Frame : 2 Throws

1-9 Frames :
max throws 2
max score 10

10th Frame (special case):
max throws 3 (if 1st throw is Strike)
max score 30

valid Frame input [0-10]

 */

class TenPinBall
{

    /**
     * @var i/p's
     */
    protected $gameData;
    /**
     * @var array contains errors if any
     */
    public $errorArray = [];
    /**
     * @var array result array
     */
    public $resultArray = [];

    /**
     * TenPinBall constructor.
     * @param $gameData
     */
    function __construct($gameData)
{
    $this->gameData = $gameData;
    unset($errorArray);
    unset($resultArray);
}

    /**
     * @return array|mixed
     */
    function getResults(){
    $this->validateInput();

    if(isset($this->errorArray[0])){
        $err =  $this->errorArray[0];
        unset($this->errorArray[0]); // Empty array for next inputs
        return $err;


    }
    else {
        $index=0;
        foreach ($this->gameData as $frame) {
            array_push($this->resultArray,$this->calculateFrameResult($frame,$index));
            $index = $index + 1;
        }
       return $this->resultArray;
    }
}

    /**
     * @param $frame
     * @param $index
     * @return int
     */
    private function calculateFrameResult($frame, $index)
{
    if (sizeof($this->resultArray) == 0) {
        $lastElement = 0;
    } else {
        $lastElement = end($this->resultArray);
    }

    if(sizeof($frame)== 1){
        return ($lastElement + $this->calculateStrike(array_sum($frame), $index,0));
    } else {
        return $lastElement + array_sum($frame);
    }
}

    /**
     * @param $sFrame
     * @param $sIndex
     * @param $stmp
     * @return int
     */
    private function calculateStrike($sFrame, $sIndex, $stmp){
if(sizeof($this->gameData[$sIndex+1]) == 1 ){
    $stmp = $stmp + $sFrame;
    return $this->calculateStrike(array_sum($this->gameData[$sIndex+1]),$sIndex+1, $stmp);
}
else{
    return $sFrame + array_sum($this->gameData[$sIndex+1]) + $stmp;
}
}

    /**
     * validates input conditions:
        given input is list of list,
        size of given data should be 10,
        1-9 Frame must have size 1 or 2,
        elements of Frame must be 0 <= element <=10
        addition must not exceed to 10,
        special case for 10th Frame
     */
    private function validateInput(){
    if (!is_array($this->gameData)){
            $error = "Invalid Game Data";
            array_push($this->errorArray,$error);
            return;
    }
    if(sizeof($this->gameData) != 10) {
        $error = "Invalid Game Data, Expecting 10 Frames, ".sizeof($this->gameData) ." Frames given!";
       array_push($this->errorArray,$error);
        return;
    }
    //Validate Total and Faulty Data
    for($i = 0 ; $i < sizeof($this->gameData)-1; $i++) {
        $tmparr = $this->gameData[$i];
        $min = 0;
        $max = 10;
        if (sizeof($tmparr) == 2) {
            //echo $tmparr[0];
            if (!($tmparr[0] >= $min && $tmparr[0] <= $max)) {
                $error = "sdsadasInvalid Frame Data,  " . json_encode($tmparr);
                echo (($min <= $tmparr[0]) && ($tmparr[0] <= $max));
                array_push($this->errorArray, $error);
                return;
            }
            if (!($tmparr[1] >= $min && $tmparr[1] <= $max)) {
                $error = "Invalid Frame Data,  " . json_encode($tmparr);
                array_push($this->errorArray, $error);
                return;
            }

        } elseif (sizeof($tmparr) == 1) {
            if ($tmparr[0] != 10) {
                $error = "Invalid Frame Data,  " . json_encode($tmparr);
                array_push($this->errorArray, $error);
                return;
            }


        }
        if(array_sum($tmparr) > 10 ){
            $error = "Invalid Score, Score should not exceed to 10, current score " . array_sum($tmparr);
            array_push($this->errorArray, $error);
            return;
        }
    }
    // 10th Frame
    if(sizeof($this->gameData[9]) == 3){
        list($v1, $v2, $v3) = $this->gameData[9];
        if($v1 != 10) {
            $error = "Invalid Data, " .$v1 ." Given";
            array_push($this->errorArray, $error);
            return;
        }
        if($v1 > 10 || $v2 > 10 || $v3 > 10) {
            $error = "Invalid Score, Score should not exceed to 10 each game, current score " . json_encode($this->gameData[9]);
            array_push($this->errorArray, $error);
            return;
        }

        if(array_sum($this->gameData[9]) > 30) {
            $error = "Invalid Score, Score should not exceed to 30, current score " . array_sum($this->gameData[9]);
            array_push($this->errorArray, $error);
            return;
        }
    }
    else{
        list($v1, $v2) = $this->gameData[9];
        $min = 0;
        $max = 10;

        if (!($v1 >= $min && $v1 <= $max)) {
            $error = "Invalid Frame Data,  " . $v1;
            array_push($this->errorArray, $error);
            return;
        }
        if (!($v2 >= $min && $v2 <= $max)) {
            $error = "Invalid Frame Data,  " . $v2;
            array_push($this->errorArray, $error);
            return;
        }
        }
    }
}

File: python/TenPinBall.py
"""
10 pin bowling

1 Game : 10 Frames
1 Frame : 2 Throws

1-9 Frames :
max throws 2
max score 10

10th Frame (special case):
max throws 3 (if 1st throw is Strike)
max score 30

valid Frame input [0-10]

"""


class TenPinBall:
    resultArray = []

    def __init__(self, game_data):
        self.game_data = game_data
        self.resultArray = []

    def get_results(self):
        message, result = self.validate_game_data()
        if not result:
            return message
            #exit(0)
        else:
            idx = 0
            for frame in self.game_data:
                self.resultArray.append(self.calculate_frame_result(frame, idx))
                idx = idx+1
        return self.resultArray

    def calculate_frame_result(self, frame, index):
        if len(self.resultArray) == 0:
            return sum(frame)

        if len(frame) == 1:
        # check for next variable and return it here
            return self.resultArray[-1] + self.calculate_strike(sum(frame), index)
        else:
            return self.resultArray[-1] + sum(frame)

    def calculate_strike(self, sframe, sindex, prevcalc = 0):
        """
        :param sframe: int
        :param sindex: int index of current item
        :param prevcalc: int , temp holder for consecutive strikes
        :return: sum of ith and i+1th element
        """
        if len(self.game_data[sindex+1]) == 1:
            """lookup for consecutive strike, loop until no strike found"""
            prevcalc = prevcalc + sframe
            return self.calculate_strike(sum(self.game_data[sindex+1]), sindex+1, prevcalc)

        else:
            return sframe + sum(self.game_data[sindex+1]) +prevcalc

    def validate_game_data(self,tdata=[]):
        """
        :return: False with Error message or True if given array is clean
        TODO: refactoring atm this block check for all necessay validation
        validations: given input is list of list,
                     size of given data should be 10,
                     1-9 Frame must have size 1 or 2,
                     elements of Frame must be 0 <= element <=10
                     addition must not exceed to 10,
                     special case for 10th Frame

        """
        for i in range(len(self.game_data) - 1):
            if not (isinstance(self.game_data[i], list)):
                return 'Invalid Game Data', False
        if len(self.game_data) != 10:
            return 'Invalid Game Length', False
        for i in range(len(self.game_data)-1):
            frame = self.game_data[i]

            if not (len(frame) == 2 or len(frame) == 1):
                return 'Invalid Frame data : ' + str(self.game_data[i]), False
            if len(frame) == 2:
                if not (self.game_data[i][0] in range(0, 11) and self.game_data[i][1] in range(0, 11)):
                    return "Not In range " + str(self.game_data[i]), False
            else:
                if not (self.game_data[i][0] in range(10, 11)):
                    return "Not Valid Result " + str(self.game_data[i]), False
            """Frame score should not exceed 10"""
            if sum(self.game_data[i]) > 10:
                return "Not Valid Result, maximum score of 10 is allowed,  given: " + str(self.game_data[i]), False

        """10th Frame"""
        if self.game_data[9][0] != 10 and len(self.game_data[9]) == 3:
            # boundary condition to check if 1st element is strike
            return 'Invalid Frame data : ' + str(self.game_data[9]), False
        if not (len(self.game_data[9]) in range(2, 4)):
            return 'Invalid Frame data : ' + str(self.game_data[9]), False
        if len(self.game_data[9]) == 2:
                if not (self.game_data[9][0] in range(0, 11) and self.game_data[9][1] in range(0, 11)):
                    return "Not In range" + str(self.game_data[9]), False
        if len(self.game_data[9]) == 3:
                if not (self.game_data[9][0] in range(0, 11) and self.game_data[9][1] in range(0, 11) and self.game_data[9][2] in range(0, 11)):
                    return "Not In range" + str(self.game_data[9]), False
        if sum(self.game_data[9]) > 30:
            return "Not Valid Result, maximum score of 30 is allowed " + str(self.game_data[9]), False

        return '_', True

print('\n')
print('-' * 4 + 'Valid inputs' + '-' * 4)
print('\n')
tenpinball = TenPinBall([[5,5],[8,1],[6,4],[10],[0,5],[2,6],[8,1],[5,3],[6,1],[10,2,6]])
print(tenpinball.get_results())
tenpinball = TenPinBall([[5,2],[8,1],[6,4],[10],[10],[2,6],[8,1],[5,3],[6,1],[10,2,6]])
print(tenpinball.get_results())
tenpinball = TenPinBall([[1,3],[2,5],[1,7],[10],[10],[5,3],[2,1],[1,1],[1,3],[10,2,1]])
print(tenpinball.get_results())

print('\n')
print('-' * 4 + 'Invalid inputs' + '-' * 4)
print('\n')
tenpinball = TenPinBall('test string')
print(tenpinball.get_results())
tenpinball = TenPinBall([[8,1],[6,4],[10],[10],[2,6],[8,1],[5,3],[6,1],[10,2,6]])
print(tenpinball.get_results())
tenpinball = TenPinBall([[10,5],[8, 6], [6, 4], [10], [10], [2, 6], [8, 1], [5, 3], [6, 1], [10, 2, 6]])
print(tenpinball.get_results())
tenpinball = TenPinBall([[1,5],[1, 6], [6, 4], [10], [10], [2, 6], [8, 1], [5, 3], [6, 1], [10, 12, 6]])
print(tenpinball.get_results())
tenpinball = TenPinBall([[1,5],[1, 6], [6, 4], [10], [10], [2, 6], [8, 1], [5, 3], [6, 1],  [6, 1],  [6, 1], [10, 12, 6]])
print(tenpinball.get_results())

